### DISCUSSION TITLE
`HandleErrorLayer` without timeout

### DISCUSSION THREAD

**Discussion by BratSinot** [tokio-rs/axum#1217] [Q&A] [answered]
Greetings!

Any way to create custom `error -> Response` handler without timeout?
This builds fine:
```rust
        .layer(
            ServiceBuilder::new()
                .layer(HandleErrorLayer::new(parse_reject))
                .timeout(std::time::Duration::from_secs(30))
        )
```

this give big template error:
```rust
        .layer(
            ServiceBuilder::new()
                .layer(HandleErrorLayer::new(parse_reject))
        )
```
```log
error[E0277]: the trait bound `axum::error_handling::HandleError<axum::routing::Route<_>, fn(std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>) -> impl std::future::Future<Output = (axum::http::StatusCode, axum::Json<api::ResponseError>)> {api::parse_reject}, _>: tower::Service<axum::http::Request<_>>` is not satisfied
   --> server/src/api/mod.rs:21:10
    |
21  |         .layer(
    |          ^^^^^ the trait `tower::Service<axum::http::Request<_>>` is not implemented for `axum::error_handling::HandleError<axum::routing::Route<_>, fn(std::boxed::Box<(dyn std::error::Error + std::marker::Send + std::marker::Sync + 'static)>) -> impl std::future::Future<Output = (axum::http::StatusCode, axum::Json<api::ResponseError>)> {api::parse_reject}, _>`
    |
    = help: the following other types implement trait `tower::Service<Request>`:
              axum::error_handling::HandleError<S, F, ()>
              axum::error_handling::HandleError<S, F, (T1, T2)>
              axum::error_handling::HandleError<S, F, (T1, T2, T3)>
              axum::error_handling::HandleError<S, F, (T1, T2, T3, T4)>
              axum::error_handling::HandleError<S, F, (T1, T2, T3, T4, T5)>
              axum::error_handling::HandleError<S, F, (T1, T2, T3, T4, T5, T6)>
              axum::error_handling::HandleError<S, F, (T1, T2, T3, T4, T5, T6, T7)>
              axum::error_handling::HandleError<S, F, (T1, T2, T3, T4, T5, T6, T7, T8)>
            and 9 others
note: required by a bound in `axum::Router::<B>::layer`
   --> /Users/asinotov/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-0.5.13/src/routing/mod.rs:297:21
    |
297 |         L::Service: Service<Request<NewReqBody>, Response = Response<NewResBody>, Error = Infallible>
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `axum::Router::<B>::layer`
```

**Comment by jplatte** [comment]
Without a layer like the timeout one, there are no errors to handle. Have you read [the error handling documentation](https://docs.rs/axum/latest/axum/error_handling/index.html)?

**Reply by BratSinot** [reply]
> Have you read [the error handling documentation](https://docs.rs/axum/latest/axum/error_handling/index.html)?

Yep, but it's not very intuitive.

I'm trying to implement custom `error -> Response` handler. From what I got, I can't make some middleware to handle all possible errors (like recover filter in warp: https://docs.rs/warp/latest/warp/trait.Filter.html#method.recover).

So, I found what I can place extractor in `Result` and use some free function for that purpose. But I stuck with extractors:
```rust
#[inline(always)]
pub(crate) async fn handle(
    _: HandlePath, // implement TypedPath
    context: Context,
    param: Result<ContentLengthLimit<Query<Param>, JSON_BODY_SIZE_LIMIT>, JsonRejection>,
) -> Result<Json<ResponseOk>, (StatusCode, Json<ResponseError>)> {

<...>

.typed_post(move |path, param| handle(path, context, param))
```

and got compiler errors:
```log
error[E0277]: the trait bound `[closure@server/src/api/mod.rs:16:21: 16:82]: axum::handler::Handler<_, _>` is not satisfied
   --> server/src/api/mod.rs:16:10
    |
16  |         .typed_post(move |path, param| update_filter_handle(path, context, param))
    |          ^^^^^^^^^^ the trait `axum::handler::Handler<_, _>` is not implemented for `[closure@server/src/api/mod.rs:16:21: 16:82]`
    |
    = help: the trait `axum::handler::Handler<T, ReqBody>` is implemented for `axum::handler::Layered<S, T>`
note: required by a bound in `axum_extra::routing::RouterExt::typed_post`
   --> /Users/asinotov/.cargo/registry/src/github.com-1ecc6299db9ec823/axum-extra-0.3.6/src/routing/mod.rs:100:12
    |
100 |         H: Handler<T, B>,
    |            ^^^^^^^^^^^^^ required by this bound in `axum_extra::routing::RouterExt::typed_post`
```

**Reply by jplatte** [reply]
If you get an error about `Handler` not being implemented, have a look at [debug_handler](https://docs.rs/axum/latest/axum/attr.debug_handler.html) which helps pinpoint the exact problem.

For customizing rejections from extractors, have a look at the [customize-extractor-error example](https://github.com/tokio-rs/axum/blob/79a0a54bc9f0f585c974b5e6793541baff980662/examples/customize-extractor-error/src/main.rs).

**Reply by BratSinot** [reply]
> For customizing rejections from extractors, have a look at the [customize-extractor-error example](https://github.com/tokio-rs/axum/blob/79a0a54bc9f0f585c974b5e6793541baff980662/examples/customize-extractor-error/src/main.rs).

I saw that, but implementing your own extract (in my case it `ContentLengthLimit` and `Query`) seems strange way for me =)
But ok, at least writing wrappers is better than nothing =)

> If you get an error about Handler not being implemented, have a look at [debug_handler](https://docs.rs/axum/latest/axum/attr.debug_handler.html) which helps pinpoint the exact problem.
That working in case, when you use `Extension`. I use closures.

**Reply by davidpdrsn** [reply]
You don't "handle" errors in axum. You implement `IntoResponse` for your error types and they'll automatically be converted into responses and sent the client.

See the error handling example the repo for a demo.

This probably something we can make clearer in the docs. Been getting this question quite often lately.

**Reply by BratSinot** [reply]
> Been getting this question quite often lately.

Because people start using `axum` =)
And, of course, they want to have one error view.

**Reply by davidpdrsn** [reply]
> And, of course, they want to have one error view.

That's not a given. With axum's model you're guaranteed that all errors are handled appropriately. Warps "recover" systems requires you to _remember_ to handle all possible errors, which inevitably means you'll forget some.

**Comment by BratSinot** [comment, marked as answer]
> > For customizing rejections from extractors, have a look at the [customize-extractor-error example](https://github.com/tokio-rs/axum/blob/79a0a54bc9f0f585c974b5e6793541baff980662/examples/customize-extractor-error/src/main.rs).
> 
> I saw that, but implementing your own extract (in my case it `ContentLengthLimit` and `Query`) seems strange way for me =) But ok, at least writing wrappers is better than nothing =)
